### DISCUSSION TITLE
Authenticatable projection

### DISCUSSION THREAD

**Discussion by keizah7** [spatie/laravel-event-sourcing#372] [General]
How I can achieve authenticable projection?

I made guard for projection model `Consultant` and to laravel auth works it needs to extend 

```
use Illuminate\Foundation\Auth\User as Authenticatable;
```

But projection also is abstract class.

```
use Spatie\EventSourcing\Projections\Projection;
```

https://laracasts.com/discuss/channels/laravel/make-model-authenticable-when-it-is-extending-another-class

**Comment by sebastiandedeyne** [comment]
Unfortunately there's no solution to extend two classes.

I suggest reimplementing Laravel's `User` class in your app. The `User` class doesn't have any specific implementations but collects a few interfaces & traits, so you don't need to worry about it getting out of date when Laravel makes changes in the underlying code.

https://github.com/laravel/framework/blob/345a2eadeb0bec3f98000e55a3ba8b304242c93e/src/Illuminate/Foundation/Auth/User.php

**Reply by keizah7** [reply]
I know that php doesn't support multiple inheritance.

Problem is that core laravel code and package have type check of classes, so traits can't solve this issue. I can move functionality, but some methods still  expects argument type of `Illuminate\Contracts\Auth\Authenticatable` or `Spatie\EventSourcing\Projections\Projection`

Thanks, I will try

**Comment by SimonMacIntyre** [comment]
Do you think something like this might fit in this package? This is more-or-less what you suggested right?

```php
<?php

namespace App\Support;

use Illuminate\Auth\Authenticatable;
use Illuminate\Auth\MustVerifyEmail;
use Illuminate\Auth\Passwords\CanResetPassword;
use Spatie\EventSourcing\Projections\Projection;
use Illuminate\Contracts\Auth\Access\Authorizable as AuthorizableContract;
use Illuminate\Contracts\Auth\Authenticatable as AuthenticatableContract;
use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;
use Illuminate\Foundation\Auth\Access\Authorizable;

class AuthenticatableProjection extends Projection implements
    AuthenticatableContract,
    AuthorizableContract,
    CanResetPasswordContract
{
    use Authorizable;
    use Authenticatable;
    use MustVerifyEmail;
    use CanResetPassword;
}
```

**Reply by keizah7** [reply]
Yes, it is right, but after that I got another problem, so I have to find another solution

**Comment by keizah7** [comment]
Thank you for tips,
the only solution was create another class, and implement all needed data:
```
<?php

namespace App\Contracts;

use Illuminate\Auth\Authenticatable;
use Illuminate\Auth\MustVerifyEmail;
use Illuminate\Auth\Passwords\CanResetPassword;
use Illuminate\Contracts\Auth\Access\Authorizable as AuthorizableContract;
use Illuminate\Contracts\Auth\Authenticatable as AuthenticatableContract;
use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;
use Illuminate\Foundation\Auth\Access\Authorizable;
use Spatie\EventSourcing\Projections\Projection;

class AuthenticableProjection extends Projection implements
    AuthenticatableContract,
    AuthorizableContract,
    CanResetPasswordContract
{
    use Authenticatable;
    use Authorizable;
    use CanResetPassword;
    use MustVerifyEmail;
}
```

but unfortunately in next step I'm getting project specific exception and the only way to fix it is to rewrite authentication logic. I won't do that, instead of it I will create another model and make relationship between them
```
Spatie \ EventSourcing \ Projections \ Exceptions \ ReadonlyProjection
The `App\Projections\Consultant` projection is not writeable at this point, please call `$model->writeable()` first.
```

**Reply by SimonMacIntyre** [reply]
That is a common error, you'd have to find that code and make it invoke `writeable()` first. If it is third-party package, you'll have to extend it (just like we did here with extending `Projection`).

**Reply by keizah7** [reply]
Yeah, I checked out the code, but model is called in laravel core files, so where is a lot of job, to rewrite so much logic to invoke `writeable` there.

My solution:
```
<?php

namespace App\Models;

use Illuminate\Contracts\Auth\MustVerifyEmail;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Notifications\Notifiable;

class Consultant extends \Illuminate\Foundation\Auth\User implements MustVerifyEmail
{
    use Notifiable;

    protected $table = 'auth_consultants';

    public function user(): BelongsTo
    {
        return $this->belongsTo(\App\Projections\Consultant::class, 'consultant_uuid');
    }
}
```
